File: wrappers/api-clients/python/examples/error_handling.py
#!/usr/bin/env python3
"""Demonstrate the per-status error class hierarchy.

The SDK raises a typed subclass for every documented status code so
callers can ``except`` the specific case they care about. All
subclasses share an `APIError` base; everything that ever leaves the
SDK inherits from `CheckrdError`."""

from __future__ import annotations

import os

import checkrd_api
from checkrd_api import (
    AuthenticationError,
    Checkrd,
    NotFoundError,
    RateLimitError,
)


def main() -> None:
    client = Checkrd(api_key=os.environ.get("CHECKRD_API_KEY", "ck_live_invalid"))

    try:
        client.agents.retrieve("00000000-0000-0000-0000-000000000000")
    except AuthenticationError as e:
        print(f"401 — bad credentials. request_id={e.request_id}")
    except NotFoundError as e:
        print(f"404 — agent does not exist in this workspace. message={e.message}")
    except RateLimitError as e:
        print(f"429 — slow down. retry after: {e.response.headers.get('retry-after')}s")
    except checkrd_api.APIStatusError as e:
        print(f"unexpected status {e.status_code}: {e.message}")
    except checkrd_api.APIConnectionError:
        print("could not reach the control plane (DNS, TCP, TLS)")


if __name__ == "__main__":
    main()

File: wrappers/api-clients/javascript/examples/error-handling.ts
/**
 * Demonstrate the per-status error class hierarchy.
 *
 * Run with: npx tsx examples/error-handling.ts
 */
import Checkrd, {
  AuthenticationError,
  APIConnectionError,
  APIStatusError,
  NotFoundError,
  RateLimitError,
} from "@checkrd/api";

async function main() {
  const client = new Checkrd({
    apiKey: process.env.CHECKRD_API_KEY ?? "ck_live_invalid",
  });

  try {
    await client.agents.retrieve("00000000-0000-0000-0000-000000000000");
  } catch (err) {
    if (err instanceof AuthenticationError) {
      console.log(`401 — bad credentials. requestId=${err.requestId ?? "<none>"}`);
    } else if (err instanceof NotFoundError) {
      console.log(`404 — agent does not exist in this workspace. message=${err.message}`);
    } else if (err instanceof RateLimitError) {
      const retryAfter = err.response.headers.get("retry-after");
      console.log(`429 — slow down. retry after: ${retryAfter ?? "?"}s`);
    } else if (err instanceof APIStatusError) {
      console.log(`unexpected status ${err.status}: ${err.message}`);
    } else if (err instanceof APIConnectionError) {
      console.log("could not reach the control plane (DNS, TCP, TLS)");
    } else {
      throw err;
    }
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
